airqo-platform/AirQo-api
# [chore] data segregation

- **What does this PR do?**

Data segragation where different organisation of the AirQo product have different databases (within the same platform). An effort towards opening up the platform. This handles two microservices (`auth-service` and `device-registry`)

With this PR, every endpoint you test must have a query parameter tenant which mentions the name of the organisation whose database you intend to access.

In the next version of this process, we shall review the DB creation process so that we do not waste resources by accidentally creating many databases that are not in active usage.

- **What JIRA task is related to this PR?**

This is linked to this [JIRA task](https://airqoteam.atlassian.net/browse/PLAT-73?atlOrigin=eyJpIjoiNzZhYzY1ZTJiMGZhNDlhOWE4YzE3OTczMDc0Mzg0M2UiLCJwIjoiaiJ9).

- **How do I test out this PR?**

> first test out  (`auth service`)

Create a docker image locally: `docker build -t us.gcr.io/airqo-250220/airqo-auth-service-api .`
Afterwards, run the newly created image: `docker run -d -p 5000:3000 us.gcr.io/airqo-250220/airqo-auth-service-api`
Using your REST client, test out all the endpoints for the` auth-service`, use port 5000 as specified above or change to what you would like to use.

Alternatively:
`cd AirQo-api/src/auth-service`
Windows: `npm run dev-pc` OR `npm run stage-pc`
Mac:  `npm run dev-mac` OR `npm run stage-mac`

> then afterwards test out (`device-registry`) 

Create a docker image locally: `docker build -t us.gcr.io/airqo-250220/airqo-device-registry-api .`
Afterwards, run the newly created image: `docker run -d -p 6000:3000 us.gcr.io/airqo-250220/airqo-device-registry-api`
Using your REST client, test out all the endpoints for the` device-registry`, use port 6000 as specified above or change to what you would like to use.

Alternatively:
`cd AirQo-api/src/device-registry`
Windows: `npm run dev-pc` OR `npm run stage-pc`
Mac:  `npm run dev-mac` OR `npm run stage-mac`


- **Which end points should are ready for testing?**

All endpoints for `auth-service` and `device-registry`

> Example:

 
http://localhost:3000/api/v1/users/addWithTenant?tenant={ORGANISATION_NAME} where ORGANISATION NAME can be any organisation -----  a new database will be created for that organisation name you have specified. 

Example of new DB name where `ORGANISTION NAME` in the URL is `kcca`: `airqo-auth-service_kcca`.Check your DB environment to confirm any new DBs created for new organisations. For an existing organisation, a new DB is NOT created but rather just modified accordingly depending on the operation.

- **Any other considerations?**

In this [API documentation](https://docs.google.com/spreadsheets/d/1-NGTDBtTcIi8Fg2UYSWjcK6geKrZdopz5kxRFsCEMPs/edit?usp=sharing), please use queries instead of parameters for all endpoints when running the two services (`auth-service` and `device-registry`) under this branch. Example: [DELETE USER] /api/v1/users/{user ID} changes to: /api/v1/users?id={USER ID}

Node environment used is v12.18.1

In case you are having challenges with environment or missing package installations - even after running npm install, please delete the node_modules folder and rerun `npm install `
